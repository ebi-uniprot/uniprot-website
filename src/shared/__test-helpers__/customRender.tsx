/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
