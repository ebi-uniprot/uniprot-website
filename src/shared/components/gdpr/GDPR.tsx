import { FC } from 'react';
import { Link } from 'react-router-dom';
import { Button } from 'franklin-sites';

import { useUserPreference } from '../../contexts/UserPreferences';

import './styles/gdpr.scss';

const GDPR: FC = () => {
  const [token, setToken] = useUserPreference<boolean>('gdpr', false);

  if (token === true) {
    return null;
  }

  return (
    <div className="gdpr-section">
      {`We'd like to inform you that we have updated our `}
      <Link to="/help/privacy">Privacy Notice</Link>
      {` to
      comply with Europeâ€™s new General Data Protection Regulation (GDPR) that
      applies since 25 May 2018.`}
      <Button variant="secondary" onClick={() => setToken(true)}>
        Accept
      </Button>
    </div>
  );
};

export default GDPR;
