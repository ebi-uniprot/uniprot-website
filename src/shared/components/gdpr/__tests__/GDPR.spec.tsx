import { screen, fireEvent, waitFor } from '@testing-library/react';
import { frame } from 'timing-functions';

import customRender from '../../../__test-helpers__/customRender';

import GDPR from '../GDPR';

describe('GDPR', () => {
  test('should render', () => {
    const { asFragment } = customRender(<GDPR />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should add 'gdpr': true to localStorage", async () => {
    expect(localStorage.getItem('gdpr')).toBe(null);
    customRender(<GDPR />);
    // Needed to wait for the user preference to be initialised
    await waitFor(frame);
    expect(
      screen.getByRole('link', { name: 'Privacy Notice' })
    ).toBeInTheDocument();
    const acceptButton = screen.getByRole('button', { name: 'Accept' });
    fireEvent.click(acceptButton);
    expect(
      screen.queryByRole('link', { name: 'Privacy Notice' })
    ).not.toBeInTheDocument();
    await waitFor(() => {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(JSON.parse(window.localStorage.getItem('gdpr')!)).toBe(true);
    });
  });

  test("if 'gdpr' in localStorage, do not render component", async () => {
    localStorage.setItem('gdpr', 'true');
    customRender(<GDPR />);
    // Since we had to wait in the other it's to be sure it doesn't appear later
    await waitFor(frame);
    expect(
      screen.queryByRole('link', { name: 'Privacy Notice' })
    ).not.toBeInTheDocument();
  });
});
