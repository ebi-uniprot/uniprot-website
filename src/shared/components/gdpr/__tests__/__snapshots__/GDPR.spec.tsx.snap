// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GDPR should render 1`] = `
<DocumentFragment>
  <div
    class="gdpr-section"
  >
    <p>
      We'd like to inform you that we have updated our 
      <a
        href="/help/privacy"
      >
        Privacy Notice
      </a>
       to comply with Europe's General Data Protection Regulation (GDPR) that applies since 25 May 2018.
    </p>
    <p>
      UniProt uses Google Analytics, Hotjar and Sentry as third-party analytics services to collect information about website performance and how users navigate through and use our sites to help us design better interfaces. We do not use any of these services to track you individually or collect personal data.
    </p>
    <button
      class="button secondary"
      type="button"
    >
      Accept
    </button>
  </div>
</DocumentFragment>
`;
